File: surface/widgets/compose/components/widgets/form-controls/ProgressBar.kt
// ============================================================
// Clef Surface Compose Widget — ProgressBar
//
// Visual progress indicator. Renders a Material 3
// LinearProgressIndicator with an optional label and
// percentage readout. Maps the progress-bar.widget anatomy
// (root, track, fill, label, valueText) to Compose layout
// with LinearProgressIndicator.
// ============================================================

package clef.surface.compose.components.widgets.formcontrols

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

enum class ProgressBarSize { Sm, Md, Lg }

// --------------- Size Mapping ---------------

private fun ProgressBarSize.trackHeight(): Dp = when (this) {
    ProgressBarSize.Sm -> 4.dp
    ProgressBarSize.Md -> 8.dp
    ProgressBarSize.Lg -> 12.dp
}

// --------------- Component ---------------

@Composable
fun ProgressBar(
    value: Float,
    modifier: Modifier = Modifier,
    max: Float = 100f,
    size: ProgressBarSize = ProgressBarSize.Md,
    label: String? = null,
    showValue: Boolean = true,
    color: Color = MaterialTheme.colorScheme.primary,
    trackColor: Color = MaterialTheme.colorScheme.surfaceVariant,
) {
    val clamped = value.coerceIn(0f, max)
    val progress = if (max > 0f) clamped / max else 0f
    val percent = (progress * 100).toInt()

    Column(modifier = modifier) {
        if (label != null) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelLarge,
            )
            Spacer(modifier = Modifier.height(4.dp))
        }

        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth(),
        ) {
            LinearProgressIndicator(
                progress = { progress },
                modifier = Modifier
                    .weight(1f)
                    .height(size.trackHeight()),
                color = color,
                trackColor = trackColor,
                strokeCap = StrokeCap.Round,
            )

            if (showValue) {
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = "$percent%",
                    style = MaterialTheme.typography.bodySmall,
                )
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/form-controls/ProgressBar.swift
// ============================================================
// Clef Surface SwiftUI Widget — ProgressBar
//
// Visual progress indicator. Renders a ProgressView with an
// optional label and percentage readout.
// ============================================================

import SwiftUI

// --------------- Types ---------------

enum ProgressBarSize {
    case sm, md, lg

    var trackHeight: CGFloat {
        switch self {
        case .sm: return 4
        case .md: return 8
        case .lg: return 12
        }
    }
}

// --------------- Component ---------------

/// ProgressBar view with an optional label and percentage display.
///
/// - Parameters:
///   - value: Current progress value.
///   - max: Maximum value (default 100).
///   - size: Size controlling track height.
///   - label: Optional label text.
///   - showValue: Whether to show the percentage.
///   - color: Fill color for the progress track.
///   - trackColor: Background track color.
struct ProgressBarView: View {
    var value: Float
    var max: Float = 100
    var size: ProgressBarSize = .md
    var label: String? = nil
    var showValue: Bool = true
    var color: Color = .accentColor
    var trackColor: Color = Color(.systemGray5)

    private var clamped: Float { Swift.min(Swift.max(value, 0), max) }
    private var progress: Float { max > 0 ? clamped / max : 0 }
    private var percent: Int { Int(progress * 100) }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let label = label {
                Text(label)
                    .font(.headline)
            }

            HStack(spacing: 8) {
                GeometryReader { geometry in
                    ZStack(alignment: .leading) {
                        RoundedRectangle(cornerRadius: size.trackHeight / 2)
                            .fill(trackColor)
                            .frame(height: size.trackHeight)

                        RoundedRectangle(cornerRadius: size.trackHeight / 2)
                            .fill(color)
                            .frame(width: geometry.size.width * CGFloat(progress), height: size.trackHeight)
                    }
                }
                .frame(height: size.trackHeight)

                if showValue {
                    Text("\(percent)%")
                        .font(.caption)
                }
            }
        }
        .accessibilityValue("\(percent) percent")
    }
}
